WalletWasabi/Services/WalletService.cs

```diff
@@ -650,14 +650,23 @@ private async Task<bool> ProcessTransactionAsync(SmartTransaction tx)
 						continue;
 					}
 
-					SmartCoin newCoin = new SmartCoin(txId, i, output.ScriptPubKey, output.Value, tx.Transaction.Inputs.ToTxoRefs().ToArray(), tx.Height, tx.IsRBF, anonset, foundKey.Label, spenderTransactionId: null, false, pubKey: foundKey); // Do not inherit locked status from key, that's different.
-																																																												   // If we did not have it.
+					var receiveKeys = KeyManager.GetKeys(x => tx.Transaction.Outputs.Any(y => y.ScriptPubKey == x.P2wpkhScript));
```

Comment:
Concept ACK. The approach makes sense to me.

Implementation NACK. I didn't measure this code performance but I think it could cause some trouble if you have to recover a wallet with many coinjoins because it performs a lot of operations (even when sometimes they are not strictly necessary) and these operations are in a for-loop.

I would move the code out of the foreach loop and short the cirtuits. Something like this:

```c#
bool isLikelyCoinJoinOutput = false;
var receiveKeys = KeyManager.GetKeys(x => tx.Transaction.Outputs.Any(y => y.ScriptPubKey == x.P2wpkhScript));
bool allReceivedInternal = receiveKeys.All(x => x.IsInternal);
if( allReceivedInternal )
{
	bool hasEqualOutputs = tx.Transaction.GetIndistinguishableOutputs(includeSingle: false).FirstOrDefault() != default;
	if( hasEqualOutputs )
	{
		// It is likely a coinjoin if the diff between receive and sent amount is small and have at least 2 equal outputs.
		Money spentAmount = Coins.Where(x => tx.Transaction.Inputs.Any(y => y.PrevOut == x.GetOutPoint())).Sum(x => x.Amount);
		Money receivedAmount = tx.Transaction.Outputs.Where(x => receiveKeys.Any(y => y.P2wpkhScript == x.ScriptPubKey)).Sum(x => x.Value);
		isLikelyCoinJoinOutput = spentAmount.Almost(receivedAmount, Money.Coins(0.005m));
	}
}
```

Reply:
I think there's no performance worry here. This only runs when coin is found. Our bottleneck is to fetch the blocks and transactions, to find these coins, not to do operations with them.

Reply:
I don't say there is a bottleneck here, what I say is that there is no reason to run this code for each received coin when we can simply move it out of the for loop and make sure it runs only once because a transaction (especially coinjoin transactions) can have more than two coins  (0.1, 0.2, ..., 0.x, change) for the wallet and in that case the code is executed more than once. 

My second point is similar: in case not all the received are `internal` it makes no sense to continue with the heuristics because it is clear that the transaction is not a wasabi coinjoin.

Reply:
Yes, makes perfect sense. Sorry for not understanding the first time: https://github.com/zkSNACKs/WalletWasabi/pull/2047/commits/8163e2a168883eacab8a3641c9680a4a5a550033